Keep pcap loop alive even if errors occur.

maffoo opened the pull request:
Fixes #182 

I think this is probably worth having even if the firewall configuration changes fix the pcap errors. We will want to monitor the logs for any errors that crop up in case this happens frequently, but that's a whole other thing...

pomalley commented:
Haven't tested it out yet but the code changes LGTM.

maffoo commented:
@pomalley, could you try this on Vince and ping the thread again with your results? It might be nice to turn the firewall back on (in case that's the cause of the issue) and run this to verify that you see the error get logged and that the server continues to run (and scans work, etc.).

pomalley commented:
The other one just crashed again so I'm running this branch on vince currently. Will check back in later...

pomalley commented:
This branch does... not work. When it crashed it entered in an infinite loop and kept spewing stuff to the console. Eventually I was able to kill it with ctrl+c or ctrl+z and here is what I got:

```
65) ~[classes/:na]
15:45:12.618 [Thread-1] INFO  org.pcap4j.core.PcapHandle - Starting loop.
15:45:12.618 [Thread-1] ERROR o.labrad.ethernet.server.EthernetBus - error in li
sten thread for bus \Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}
org.pcap4j.core.PcapNativeException: Error occured: read error: PacketReceivePac
ket failed
        at org.pcap4j.core.PcapHandle.doLoop(PcapHandle.java:809) ~[pcap4j-core-
1.4.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:778) ~[pcap4j-core-1.
4.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:743) ~[pcap4j-core-1.
4.0.jar:na]
        at org.labrad.ethernet.server.EthernetBus$$anon$1.run(EthernetBus.scala:
65) ~[classes/:na]
15:45:12.618 [Thread-1] INFO  org.pcap4j.core.PcapHandle - Starting loop.
15:45:12.618 [Thread-1] ERROR o.labrad.ethernet.server.EthernetBus - error in li
sten thread for bus \Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}
org.pcap4j.core.PcapNativeException: Error occured: read error: PacketReceivePac
ket failed
        at org.pcap4j.core.PcapHandle.doLoop(PcapHandle.java:809) ~[pcap4j-core-
1.4.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:778) ~[pcap4j-core-1.
4.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:743) ~[pcap4j-core-1.
4.0.jar:na]
        at org.labrad.ethernet.server.EthernetBus$$anon$1.run(EthernetBus.scala:
65) ~[classes/:na]
15:45:12.628 [Thread-1] INFO  org.pcap4j.core.PcapHandle - Starting loop.
15:45:12.628 [Thread-1] ERROR o.labrad.ethernet.server.EthernetBus - error in li
sten thread for bus \Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}
org.pcap4j.core.PcapNativeException: Error occured: read error: PacketReceivePac
ket failed
        at org.pcap4j.core.PcapHandle.doLoop(PcapHandle.java:809) ~[pcap4j-core-
1.4.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:778) ~[pcap4j-core-1.
4.0.jar:na]
```

maffoo commented:
Thanks for trying it out; I'm guessing after an error the pcap handle is invalid, so we'd need to open a new handle in order to continue.

maffoo commented:
@pomalley, I changed the logic to not try and reuse a pcap handle after an error. Also introduced a delay so that if we do get in an error loop at least we won't peg the CPU. I'd like to try this just to see whether these errors are recoverable or not; when you have a chance, could you give it a spin?

pomalley commented:
No luck; here's the stack trace this time:

```
redheatuser@REDHEATDOS /c/labrad-servers/servers/DirectEthernet (u/maffoo/de-pca
p-loop)
$ sbt "runMain org.labrad.ethernet.server.DirectEthernet"
[info] Loading project definition from C:\labrad-servers\servers\DirectEthernet\
project
[info] Set current project to DirectEthernet (in build file:/C:/labrad-servers/s
ervers/DirectEthernet/)
[info] Running org.labrad.ethernet.server.DirectEthernet
13:10:45.018 [run-main-0] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/p
cap4j.properties] Could not get value by org.pcap4j.af.inet, use default value:
null
13:10:45.018 [run-main-0] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/p
cap4j.properties] Could not get value by org.pcap4j.af.inet6, use default value:
 null
13:10:45.018 [run-main-0] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/p
cap4j.properties] Could not get value by org.pcap4j.af.packet, use default value
: null
13:10:45.028 [run-main-0] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/p
cap4j.properties] Could not get value by org.pcap4j.af.link, use default value:
null
13:10:45.028 [run-main-0] INFO  org.pcap4j.core.Pcaps - 4 NIF(s) found.
13:10:45.038 [run-main-0] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/p
cap4j.properties] Could not get value by org.pcap4j.dlt.raw, use default value:
null
13:10:45.048 [run-main-0] INFO  org.labrad.ServerConnection$ - LABRADNODE -> Vin
ce
13:10:45.478 [run-main-0] INFO  org.labrad.ServerConnection - Now serving...
13:10:47.816 [Thread-1] INFO  org.pcap4j.core.PcapHandle - Starting loop.
13:10:47.823 [Thread-1] WARN  o.p.packet.factory.PacketFactories - java.lang.NoC
lassDefFoundError:org/pcap4j/packet/factory/PacketFactoryBinder
15:54:01.689 [Thread-1] ERROR o.labrad.ethernet.server.EthernetBus - error in li
sten thread for bus \Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}
org.pcap4j.core.PcapNativeException: Error occured: read error: PacketReceivePac
ket failed
        at org.pcap4j.core.PcapHandle.doLoop(PcapHandle.java:809) ~[pcap4j-core-
1.4.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:778) ~[pcap4j-core-1.
4.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:743) ~[pcap4j-core-1.
4.0.jar:na]
        at org.labrad.ethernet.server.EthernetBus$$anon$1.run(EthernetBus.scala:
65) ~[classes/:na]
15:54:02.699 [Thread-1] INFO  org.pcap4j.core.PcapHandle - Closed.
[error] (Thread-1) org.pcap4j.core.PcapNativeException: Error opening adapter: T
he handle is invalid. (6)
org.pcap4j.core.PcapNativeException: Error opening adapter: The handle is invali
d. (6)
        at org.pcap4j.core.PcapNetworkInterface.openLive(PcapNetworkInterface.ja
va:233)
        at org.labrad.ethernet.server.EthernetBus$$anon$1.run(EthernetBus.scala:
63)
[trace] Stack trace suppressed: run last compile:runMain for the full output.
[success] Total time: 9916 s, completed Jun 10, 2015 3:56:00 PM

redheatuser@REDHEATDOS /c/labrad-servers/servers/DirectEthernet (u/maffoo/de-pca
p-loop)
$
```

maffoo commented:
Some more info. @pomalley and I added some failure handling logic to try to reopen a new pcap session after an error like this, however, we found that the handle to the actual NIF device was invalid, and in fact pcap reported not finding any devices at all after the error. As though all the network cards suddenly disappeared. This is very strange.
```
C:\labrad-servers\servers\DirectEthernet\target\pack>java -cp lib\* org.labrad.ethernet.server.DirectEthernet
12:03:44.222 [main] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/pcap4j.properties] Could not get value by org.pcap4j.af.inet, use default value: null
12:03:44.222 [main] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/pcap4j.properties] Could not get value by org.pcap4j.af.inet6, use default value: null
12:03:44.222 [main] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/pcap4j.properties] Could not get value by org.pcap4j.af.packet, use default value: null
12:03:44.222 [main] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/pcap4j.properties] Could not get value by org.pcap4j.af.link, use default value: null
12:03:44.232 [main] INFO  org.pcap4j.core.Pcaps - 4 NIF(s) found.
12:03:44.232 [main] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/pcap4j.properties] Could not get value by org.pcap4j.dlt.raw, use default value: null
12:03:44.242 [main] INFO  org.labrad.ServerConnection$ - LABRADNODE -> Vince
12:03:44.592 [main] INFO  org.labrad.ServerConnection - Now serving...
12:04:04.407 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] INFO  org.pcap4j.core.PcapHandle - Starting loop.
12:04:04.417 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] WARN  o.p.packet.factory.PacketFactories - java.lang.NoClassDefFoundError:org/pcap4j/packet/factory/PacketFactoryBinder
12:37:16.409 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] ERROR o.labrad.ethernet.server.EthernetBus - error in listen thread for bus \Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}
org.pcap4j.core.PcapNativeException: Error occured: read error: PacketReceivePacket failed
        at org.pcap4j.core.PcapHandle.doLoop(PcapHandle.java:813) ~[pcap4j-core-1.5.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:782) ~[pcap4j-core-1.5.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:747) ~[pcap4j-core-1.5.0.jar:na]
        at org.labrad.ethernet.server.EthernetBus$$anon$1.run(EthernetBus.scala:73) ~[directethernet_2.11-1.1.0.jar:1.1.0]

12:38:24.500 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] INFO  org.pcap4j.core.PcapHandle - Closed.
12:38:24.580 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] ERROR o.labrad.ethernet.server.EthernetBus - failed to open adapter name: [\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] description: [Intel(R) PRO/1000 GT Desktop Adapter] address: [/fe80:0:0:0:8ccd:fba3:7ecb:90d0] address: [/0.0.0.0] loopBack: [false]] local: [true]
org.pcap4j.core.PcapNativeException: Error opening adapter: The handle is invalid. (6)
        at org.pcap4j.core.PcapNetworkInterface.openLive(PcapNetworkInterface.java:247) ~[pcap4j-core-1.5.0.jar:na]
        at org.labrad.ethernet.server.EthernetBus$$anon$1.run(EthernetBus.scala:64) ~[directethernet_2.11-1.1.0.jar:1.1.0]
12:38:24.580 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] INFO  org.pcap4j.core.Pcaps - No NIF was found.
12:38:24.580 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] ERROR o.labrad.ethernet.server.EthernetBus - available adapters:
Exception in thread "EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}" org.pcap4j.core.PcapNativeException: Error opening adapter: The handle is invalid. (6)
        at org.pcap4j.core.PcapNetworkInterface.openLive(PcapNetworkInterface.java:247)
        at org.labrad.ethernet.server.EthernetBus$$anon$1.run(EthernetBus.scala:64)
```

maffoo commented:
Added logic to try reconnecting after pcap failure. This works, provided we get a new handle to the pcap device. It seems this is able to work and restart the packet receive loop, though it may take a few retries:
```
C:\labrad-servers\servers\DirectEthernet\target\pack>java -cp lib\* org.labrad.e
thernet.server.DirectEthernet
22:49:30.335 [main] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/pcap4j.
properties] Could not get value by org.pcap4j.af.inet, use default value: null
22:49:30.335 [main] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/pcap4j.
properties] Could not get value by org.pcap4j.af.inet6, use default value: null
22:49:30.335 [main] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/pcap4j.
properties] Could not get value by org.pcap4j.af.packet, use default value: null

22:49:30.335 [main] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/pcap4j.
properties] Could not get value by org.pcap4j.af.link, use default value: null
22:49:30.335 [main] INFO  org.pcap4j.core.Pcaps - 4 NIF(s) found.
22:49:30.345 [main] WARN  org.pcap4j.util.PropertiesLoader - [org/pcap4j/pcap4j.
properties] Could not get value by org.pcap4j.dlt.raw, use default value: null
22:49:30.355 [main] INFO  org.labrad.ServerConnection$ - LABRADNODE -> Vince
22:49:30.685 [main] INFO  org.labrad.ServerConnection - Now serving...
22:49:30.909 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.PcapHandle - Starting loop.
22:49:30.919 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] WARN  o.p.packet.factory.PacketFactories - java.lang.NoClassDefFoundE
rror:org/pcap4j/packet/factory/PacketFactoryBinder
23:28:31.316 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] ERROR o.labrad.ethernet.server.EthernetBus - error in listen thread f
or bus \Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}
org.pcap4j.core.PcapNativeException: Error occured: read error: PacketReceivePac
ket failed
        at org.pcap4j.core.PcapHandle.doLoop(PcapHandle.java:813) ~[pcap4j-core-
1.5.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:782) ~[pcap4j-core-1.
5.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:747) ~[pcap4j-core-1.
5.0.jar:na]
        at org.labrad.ethernet.server.EthernetBus$$anon$1.run(EthernetBus.scala:
74) ~[directethernet_2.11-1.1.0.jar:1.1.0]
23:29:01.510 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.Pcaps - No NIF was found.
23:29:01.510 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  o.labrad.ethernet.server.EthernetBus - attempt=1. device=null
23:29:02.551 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.Pcaps - 4 NIF(s) found.
23:29:02.551 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  o.labrad.ethernet.server.EthernetBus - attempt=1. device=name:
[\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] description: [Intel(R) PRO/
1000 GT Desktop Adapter] loopBack: [false]] local: [true]
23:29:02.551 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.PcapHandle - Closed.
23:29:02.551 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.PcapHandle - Starting loop.
23:29:02.701 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] ERROR o.labrad.ethernet.server.EthernetBus - error in listen thread f
or bus \Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}
org.pcap4j.core.PcapNativeException: Error occured: read error: PacketReceivePac
ket failed
        at org.pcap4j.core.PcapHandle.doLoop(PcapHandle.java:813) ~[pcap4j-core-
1.5.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:782) ~[pcap4j-core-1.
5.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:747) ~[pcap4j-core-1.
5.0.jar:na]
        at org.labrad.ethernet.server.EthernetBus$$anon$1.run(EthernetBus.scala:
74) ~[directethernet_2.11-1.1.0.jar:1.1.0]
23:29:03.741 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.Pcaps - 4 NIF(s) found.
23:29:03.741 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  o.labrad.ethernet.server.EthernetBus - attempt=1. device=name:
[\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] description: [Intel(R) PRO/
1000 GT Desktop Adapter] address: [/fe80:0:0:0:8ccd:fba3:7ecb:90d0] loopBack: [f
alse]] local: [true]
23:29:03.741 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.PcapHandle - Closed.
23:29:03.741 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.PcapHandle - Starting loop.
23:59:01.220 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] ERROR o.labrad.ethernet.server.EthernetBus - error in listen thread f
or bus \Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}
org.pcap4j.core.PcapNativeException: Error occured: read error: PacketReceivePac
ket failed
        at org.pcap4j.core.PcapHandle.doLoop(PcapHandle.java:813) ~[pcap4j-core-
1.5.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:782) ~[pcap4j-core-1.
5.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:747) ~[pcap4j-core-1.
5.0.jar:na]
        at org.labrad.ethernet.server.EthernetBus$$anon$1.run(EthernetBus.scala:
74) ~[directethernet_2.11-1.1.0.jar:1.1.0]
23:59:26.510 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.Pcaps - No NIF was found.
23:59:26.510 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  o.labrad.ethernet.server.EthernetBus - attempt=1. device=null
23:59:27.552 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.Pcaps - 4 NIF(s) found.
23:59:27.552 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  o.labrad.ethernet.server.EthernetBus - attempt=1. device=name:
[\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] description: [Intel(R) PRO/
1000 GT Desktop Adapter] loopBack: [false]] local: [true]
23:59:27.552 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.PcapHandle - Closed.
23:59:27.552 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.PcapHandle - Starting loop.
23:59:27.572 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] ERROR o.labrad.ethernet.server.EthernetBus - error in listen thread f
or bus \Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}
org.pcap4j.core.PcapNativeException: Error occured: read error: PacketReceivePac
ket failed
        at org.pcap4j.core.PcapHandle.doLoop(PcapHandle.java:813) ~[pcap4j-core-
1.5.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:782) ~[pcap4j-core-1.
5.0.jar:na]
        at org.pcap4j.core.PcapHandle.loop(PcapHandle.java:747) ~[pcap4j-core-1.
5.0.jar:na]
        at org.labrad.ethernet.server.EthernetBus$$anon$1.run(EthernetBus.scala:
74) ~[directethernet_2.11-1.1.0.jar:1.1.0]
23:59:28.612 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.Pcaps - 4 NIF(s) found.
23:59:28.612 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  o.labrad.ethernet.server.EthernetBus - attempt=1. device=name:
[\Device\NPF_{2C339376-DD48-493F-A338-54725A07704D}] description: [Intel(R) PRO/
1000 GT Desktop Adapter] address: [/fe80:0:0:0:8ccd:fba3:7ecb:90d0] loopBack: [f
alse]] local: [true]
23:59:28.612 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.PcapHandle - Closed.
23:59:28.612 [EthernetBus-listenThread-\Device\NPF_{2C339376-DD48-493F-A338-5472
5A07704D}] INFO  org.pcap4j.core.PcapHandle - Starting loop.
```
However, in this case we observed that subsequent writes failed, presumably because the handle for packet sending had also become invalid. So, we need to add reconnect logic to the sending code as well.

maffoo commented:
Just heard from @JulianSKelly that this seems to have been a result of power management settings causing the network adapters to be shut down by windows. We could still keep the recovery logic here if there are other errors, but I'd be curious to know whether we've seen any other errors happen that weren't related to adapters getting shut down. Thoughts?

maffoo commented:
#182 is resolved but I'd like to get these changes in because they make the ethernet code more robust against other potential failure cases. I've squashed and rebased off master, can someone take a look now? Scala guys? @pomalley @DanielSank

DanielSank commented on DirectEthernet/src/main/pack/DirectEthernet.ini:
```diff
@@ -1,6 +1,6 @@
 [info]
 name = Scala Direct Ethernet
-version = 1.0.0-M1
```
Might be nice to have this brought in from a single point both here and in `build.sbt`.

maffoo replied:
I agree, though to do that we'd probably need to modify the build to produce the ini file from a template or something like that. I'll file an issue.

maffoo replied:
#199

DanielSank commented on DirectEthernet/src/main/scala/org/labrad/ethernet/server/EthernetBus.scala:
```diff
@@ -58,26 +59,84 @@ class EthernetBus(
+        while (alive) {
+          val pcap = reopen()
+          try {
+            pcap.loop(-1, packetListener)
```
What does this line do?

maffoo replied:
starts the pcap loop listening for packets. -1 means loop forever.

DanielSank replied:
Does this line block?

maffoo replied:
yes, nominally forever. That's why we spin up a separate listener thread for each ethernet bus.

DanielSank commented:
I'm not sure I understand the strategy. Is the idea that we're adding code such that if a send fails we try again and then die only upon multiple failures?

maffoo commented:
It's both for sending and listening for packets. If either fails, we at least attempt to recover. The particular failure that prompted this (windows putting the network adapter to sleep) turned out to be unrecoverable, but I think the recovery logic could be helpful if other (unforseen) errors occur.

DanielSank commented:
We used to catch and throw exceptions coming out of the pcap loop, but now we log errors and fall off the end of the code in the thread. I guess I just don't understand why we're making this change (probably I just don't understand the syntax...).

DanielSank commented on DirectEthernet/src/main/scala/org/labrad/ethernet/server/EthernetBus.scala:
```diff
@@ -58,26 +59,84 @@ class EthernetBus(
 
-  override def send(pkt: EthernetPacket): Unit = {
-    sender.sendPacket(pkt)
+  override def send(pkt: EthernetPacket, maxAttempts: Int = 2): Unit = {
```
Shouldn't we throw if we fail to send?

maffoo replied:
yup, it does. see comment about `sys.error` below

DanielSank commented on DirectEthernet/src/main/scala/org/labrad/ethernet/server/EthernetBus.scala:
```diff
@@ -58,26 +59,84 @@ class EthernetBus(
+      attempt += 1
+    }
+    if (pcap == null) {
+      sys.error(s"failed to reopen pcap handle in $maxAttempts attempts")
```
Maybe I'm missing something about how exceptions work in scala, but shouldn't we throw here?

maffoo replied:
`sys.error` throws an exception. `sys.error(msg)` is roughly equivalent to `throw new RuntimeException(msg)`, just shorter.

DanielSank replied:
Oh, groovy. In that case I think this LGTM.

maffoo commented:
The control flow may be a little unclear, but basically the listen thread will run forever in the while loop unless it gets interrupted (since `alive` gets set to false) or the `reopen` call fails; the thread doesn't end if other errors happen in the `pcap.loop` call. This just means if there is a (recoverable) pcap error the server will continue to run, where before it would exit after the pcap thread had an error.

DanielSank commented:
As we discussed via phonon exchange, it would be nice to have some kind of notification if there are a lot of retries. Also discussed via phonon exchange is the fact that we don't have a great system for that...
